### PR TITLE
Return empty HTTP response body when restJson1 operation has no output shape

### PR DESCRIPTION
**Pull request by david-perez**
According to the protocol tests, here's how a server should behave when
the operation has no output, and when the operation output shape is empty
(has no members):

* Empty output is serialized to `"{}"`; clients should gracefully accept `""`.
* No output is serialized to `""`.

Compare this to how clients should behave when the operation has no
input, and when the operation input shape is empty (has no members):

* Empty input is serialized to `""`; servers should gracefully accept `"{}"`.
* No input is serialized to `""`.

Notice the behavioral difference in the empty case among clients and
servers. Since clients treat both cases the same by serializing `""`,
the "no input" case is reduced to the "empty input" case by transforming
the model and creating synthetic empty input shapes for all operations
that have no input shapes.

However, the server behavior is different. Up until this commit servers
were always serializing `"{}"` in both cases. This commit implements the
correct behavior. It does so by disambiguating both cases by checking if
the operation output shape is actually synthetic, which would mean that
the original model has no operation output. In that case the structured
data serializer now returns `null`, so we end up serializing `""` to the
HTTP response body.

This commit also renames the methods in charge of serializing operation
inputs/outputs from the `StructuredDataSerializerGenerator` interface,
and removes the unused class `OutputBodyTrait` from
`SyntheticOutputTrait.kt`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by david-perez** [comment]
While this works fine, I'm not confident this is the correct fix.

I'd prefer if if we could remove the `OperationNormalizer` transform from the server plugin, and refactor the `fun OperationShape.{input,output}Shape` functions from `Smithy.kt` so that they return nullable shapes. That would be a non-negligible refactor, since all the codebase works under the assumption that operation input / output shapes always exist. However, I'm not sure I completely understand the whole purpose of `OperationNormalizer`, or if it's just a way to reduce the no input case to the empty input case for clients for convenience.

Need some guidance here @awslabs/rust-sdk-owners .

**Comment by Velfi** [comment]
This seems reasonable to me but I'm sure @rcoh will have a more informed opinion.

**Comment by crisidev** [comment]
> While this works fine, I'm not confident this is the correct fix.
> 
> I'd prefer if if we could remove the `OperationNormalizer` transform from the server plugin, and refactor the `fun OperationShape.{input,output}Shape` functions from `Smithy.kt` so that they return nullable shapes. That would be a non-negligible refactor, since all the codebase works under the assumption that operation input / output shapes always exist. However, I'm not sure I completely understand the whole purpose of `OperationNormalizer`, or if it's just a way to reduce the no input case to the empty input case for clients for convenience.
> 
> Need some guidance here @awslabs/rust-sdk-owners .

I don't think we can really remove the `OperationNormalizer`. As far as I remember it is there to generate the synthetic input / output to prevent a structure to be reused directly in multiple operations. I could be wrong, some advice from the client SDK is needed here.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/JsonSerializerGenerator.kt [hunk 1]
```diff
@@ -270,11 +272,26 @@ class JsonSerializerGenerator(
...
     }
 
-    override fun serverOutputSerializer(operationShape: OperationShape): RuntimeType {
+    override fun operationOutputSerializer(operationShape: OperationShape): RuntimeType? {
+        // Don't generate an operation JSON serializer if there was no operation output shape in the
+        // original (untransformed) model.
+        val syntheticOutputTrait = operationShape.outputShape(model).getTrait<SyntheticOutputTrait>()
+        if (syntheticOutputTrait != null && syntheticOutputTrait.originalId == null) {
+            return null
+        }
```
[the remark this answers is not in the document]
Comment (david-perez) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/JsonSerializerGenerator.kt]: Used `expectTrait` in https://github.com/awslabs/smithy-rs/pull/1208/commits/8d363cebba2d6dbda05f5bcaf75cefa02a1f4b88

**Comment by david-perez** [comment]
I'm not confident that applying the operation transformer in the server is beneficial, but I'm merging this anyway. If it comes up again we should reconsider removing it.

**Pull request merged**